lib/iris/analysis/stats.py

```diff
@@ -23,137 +23,96 @@
+        that latitude/longitude area weights can be calculated using
+        :func:`iris.analysis.cartography.area_weights`.
+    * mdtol (float, optional):
+        Tolerance of missing data. The missing data fraction is calculated
+        based on the number of grid cells masked in both cube_a and cube_b. If
+        this fraction exceed mdtol, the returned value in the corresponding
+        cell is masked. mdtol=0 means no missing data is tolerated while
+        mdtol=1 means the resulting element will be masked if and only if all
+        contributing elements are masked in cube_a or cube_b. Defaults to 1.
 
     Returns:
-        Cube of correlations.
+        A cube of the correlation between the two input cubes along the
+        specified dimensions, at each point in the remaining dimensions of the
+        cubes.
+
+        For example providing two time/altitude/latitude/longitude cubes and
+        corr_coords of 'latitude' and 'longitude' will result in a
+        time/altitude cube describing the latitude/longitude (i.e. pattern)
+        correlation at each time/altitude point.
 
     Reference:
         http://www.statsoft.com/textbook/glosp.html#Pearson%20Correlation
 
     """
 
-    # If no coords passed then set to all coords of cube.
+    # Assign larger cube to cube_1
+    if cube_b.ndim > cube_a.ndim:
+        cube_1 = cube_b
+        cube_2 = cube_a
+    else:
+        cube_1 = cube_a
+        cube_2 = cube_b
+
+    dim_coords_1 = [coord.name() for coord in cube_1.dim_coords]
+    dim_coords_2 = [coord.name() for coord in cube_2.dim_coords]
+    common_dim_coords = list(set(dim_coords_1) & set(dim_coords_2))
+    # If no coords passed then set to all common dimcoords of cubes.
     if corr_coords is None:
-        corr_coords = [c.name() for c in cube_a.dim_coords]
-
-    vec_a, vec_b, res_ind = _get_calc_view(cube_a,
-                                           cube_b,
-                                           corr_coords)
-
-    sa = vec_a - np.mean(vec_a, 0)
-    sb = vec_b - np.mean(vec_b, 0)
-    flat_corrs = np.sum((sa*sb), 0)/np.sqrt(np.sum(sa**2, 0)*np.sum(sb**2, 0))
-
-    corrs = flat_corrs.reshape([cube_a.shape[i] for i in res_ind])
-
-    # Construct cube to hold correlation results.
-    corrs_cube = iris.cube.Cube(corrs)
-    corrs_cube.long_name = "Pearson's r"
-    corrs_cube.units = "1"
-    for i, dim in enumerate(res_ind):
-        c = cube_a.dim_coords[dim]
-        corrs_cube.add_dim_coord(c, i)
-
-    return corrs_cube
+        corr_coords = common_dim_coords
+
+    # Broadcast weights to shape of cube_1 if necessary.
+    if weights is None or cube_1.shape == cube_2.shape:
+        weights_1 = weights
+        weights_2 = weights
+    else:
+        if weights.shape != cube_2.shape:
+            raise ValueError("weights array should have dimensions {}".
+                             format(cube_2.shape))
+        dims_1_common = [i for i in xrange(cube_1.ndim) if
```

Comment:
I think this needs to be just `range` rather than `xrange` for Python 3 compatibility.